File: algorithm/912.cpp
#include <vector>
#include <stdlib.h>
using namespace std;

class Solution
{
public:
    vector<int> sortArray(vector<int>& nums);

private:
    void mergeSort(vector<int>& nums, int left, int right, vector<int>& tmp);
    void merge(vector<int>& nums, int left, int mid, int right, vector<int>& tmp);
    void quickSort(vector<int>& nums, int left, int right);
    int partition(vector<int>& nums, int left, int right);
};

//选择排序。时间复杂度O(N^2)，空间复杂度O(1)
// vector<int> Solution::sortArray(vector<int>& nums)
// {
//     if(nums.size()<2) return nums;

//     int n = nums.size();
//     for (int i = 0; i < n-1; i++)
//     {
//         int min_idx = i;
//         for (int j = i+1; j < n; j++)
//         {
//             if(nums[min_idx]>nums[j])
//             {
//                 min_idx = j;
//             }
//         }
//         swap(nums[i], nums[min_idx]);
//     }
//     return nums;
// }

//冒泡排序。时间复杂度O(N^2)，空间复杂度O(1)
// vector<int> Solution::sortArray(vector<int>& nums)
// {
//     if(nums.size()<2) return nums;
//     int n = nums.size();

//     for (int i = n-1; i > 0; i--)
//     {
//         bool sorted = true;

//         for (int j = 0; j < i; j++)
//         {
//             if (nums[j]>nums[j+1])
//             {
//                 swap(nums[j],nums[j+1]);
//                 sorted = false;
//             }
//         }
//         if(sorted) break;
//     }
//     return nums;
// }

//插入排序。时间复杂度O(N^2)，空间复杂度O(1)
// vector<int> Solution::sortArray(vector<int>& nums)
// {
//     if(nums.size()<2) return nums;
//     int n = nums.size();

//     for (int i = 1; i < n; i++)
//     {
//         int temp = nums[i];
//         int j = i;
//         while (j>0 && temp<nums[j-1])
//         {
//             nums[j] = nums[j-1];
//             j--;
//         }
//         nums[j] = temp;
//     }
    
//     return nums;
// }

//归并排序。时间复杂度O(NlogN)，空间复杂度O(N)
// vector<int> Solution::sortArray(vector<int>& nums)
// {
//     if(nums.size()<2) return nums;
//     int n = nums.size();
//     vector<int> tmp(n,0);
//     mergeSort(nums, 0, n-1, tmp);
//     return nums;
// }

// void Solution::mergeSort(vector<int>& nums, int left, int right, vector<int>& tmp)
// {
//     if(left>=right) return;
//     int mid = left + ((right-left)>>1);
//     mergeSort(nums, left, mid, tmp);
//     mergeSort(nums, mid+1, right, tmp);
//     merge(nums, left, mid, right, tmp);
//     return ;
// }

// void Solution::merge(vector<int>& nums, int left, int mid, int right, vector<int>& tmp)
// {
//     if(left>=right) return;
//     if(nums[mid]<nums[mid+1]) return; //左右两区间已经有序，不用再排序

//     int i=left, j=mid+1;
//     int idx = left;
//     while (i<=mid && j<=right)
//     {
//         if(nums[i]<=nums[j]) tmp[idx++] = nums[i++];
//         else tmp[idx++] = nums[j++];
//     }
    
//     while(i<=mid) tmp[idx++] = nums[i++];
//     while(j<=right) tmp[idx++] = nums[j++];

//     for (i = left; i <= right; i++)
//     {
//         nums[i] = tmp[i];
//     }
//     return ;
// }

//快排。时间复杂度O(NlogN)，空间复杂度O(logN)
vector<int> Solution::sortArray(vector<int>& nums)
{
    if(nums.size()<2) return nums;
    int n = nums.size();
    quickSort(nums, 0, n-1);
    return nums;
}

void Solution::quickSort(vector<int>& nums, int left, int right)
{
    if(left>=right) return;
    int pivot_idx = partition(nums, left, right);
    quickSort(nums, left, pivot_idx-1);
    quickSort(nums, pivot_idx+1, right);
    return;
}

int Solution::partition(vector<int>& nums, int left, int right)
{
    int i = rand() % (right-left+1) + left;
    int pivot = nums[i];
    swap(nums[left], nums[i]);
    while (left<right)
    {
        while(left<right && nums[right]>=pivot) right--;
        nums[left] = nums[right];

        while(left<right && nums[left]<=pivot) left++;
        nums[right] = nums[left];
    }
    nums[left] = pivot;
    return left;
}

int main()
{
    vector<int> nums = {5,2,3,1};
    Solution obj;
    vector<int> result = obj.sortArray(nums);
    return 0;
}

File: algorithm/go/firstround/912.go
package firstround

import (
	"math/rand"

	"github.com/eapache/go-resiliency/breaker"
)

// 快速排序,时间复杂度O(nlogn),空间复杂度O(logn),最坏情况O(n)
func sortArray(nums []int) []int {
	if nums == nil || len(nums) < 1 {
		return nums
	}
	quickSort(nums, 0, len(nums)-1)
	return nums
}

func quickSort(nums []int, left, right int) {
	if left >= right {
		return
	}

	pivot := partition(nums, left, right)
	quickSort(nums, left, pivot-1)
	quickSort(nums, pivot+1, right)
}

func partition(nums []int, left, right int) int {
	i := rand.Intn(right-left+1) + left
	nums[left], nums[i] = nums[i], nums[left]
	pivot := left

	for left < right {
		for left < right && nums[right] >= nums[pivot] {
			right--
		}

		for left < right && nums[left] <= nums[pivot] {
			left++
		}

		nums[left], nums[right] = nums[right], nums[left]
	}
	nums[right], nums[pivot] = nums[pivot], nums[right]

	return right
}

// 堆排序,时间复杂度O(nlogn),空间复杂度O(1)
func sortArray(nums []int) []int {
	if nums == nil || len(nums) < 1 {
		return nums
	}

	heapSort(nums)
	return nums
}

func heapSort(nums []int) {
	buildHeap(nums)
	for i := len(nums) - 1; i > 0; i-- {
		nums[0], nums[i] = nums[i], nums[0]
		maxHeapify(nums, 0, i)
	}
}

func buildHeap(nums []int) {
	for i := len(nums)/2 - 1; i >= 0; i-- {
		maxHeapify(nums, i, len(nums))
	}
}

func maxHeapify(nums []int, i, len int) {
	for i <= len/2-1 {
		left, right := (i+1)*2-1, (i+1)*2

		if left >= len {
			break
		}

		maxChild := left
		if right < len && nums[right] > nums[maxChild] {
			maxChild = right
		}

		if nums[i] < nums[maxChild] {
			nums[i], nums[maxChild] = nums[maxChild], nums[i]
			i = maxChild
		} else {
			break
		}
	}
}

// 归并排序,时间复杂度O(nlogn),空间复杂度O(n)
func sortArray(nums []int) []int {
	mergeSort(nums, 0, len(nums)-1)
	return nums
}

func mergeSort(nums []int, left, right int) {
	if left >= right {
		return
	}

	mid := left + (right-left+1)/2
	mergeSort(nums, left, mid-1)
	mergeSort(nums, mid, right)
	merge(nums, left, right, mid)
}

func merge(nums []int, left, right, mid int) {
	len1, len2 := mid-left, right-mid+1
	s1 := make([]int, len1)
	s2 := make([]int, len2)
	copy(s1, nums[left:mid])
	copy(s2, nums[mid:right+1])

	i := left
	cur1, cur2 := 0, 0
	for cur1 < len1 && cur2 < len2 {
		if s1[cur1] <= s2[cur2] {
			nums[i] = s1[cur1]
			cur1++
		} else {
			nums[i] = s2[cur2]
			cur2++
		}
		i++
	}

	for cur1 < len1 {
		nums[i] = s1[cur1]
		cur1++
		i++
	}
	
	for cur2 < len2 {
		nums[i] = s2[cur2]
		cur2++
		i++
	}
}